src/coreclr/jit/morph.cpp

```diff
@@ -13163,195 +13163,94 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 // Return Value:
 //    The optimized tree (that can have any shape).
 //
-GenTree* Compiler::fgOptimizeCast(GenTree* tree)
+GenTree* Compiler::fgOptimizeCast(GenTreeCast* cast)
 {
-    GenTree*  oper    = tree->AsCast()->CastOp();
-    var_types srcType = oper->TypeGet();
-    var_types dstType = tree->CastToType();
-    unsigned  dstSize = genTypeSize(dstType);
+    GenTree* src = cast->CastOp();
```

Comment:
Are we doing these optimizations even in minopts? Should we be?

Reply:
Yes we are, that is how it has been for a very long time.

The "discard a cast" part seems low value for minopts, `optNarrowTree`, however, does a lot of work for 32 targets CQ-wise (notably, it is somewhat misleadingly under a `CLFLG_TREETRANS` guard, the only optimization enabled in minopts).

I personally think we should not be doing any optimization in minopts (which then will allow us to selectively enable high-impact optimizations in T0), since it is cheaper to have one big `opts.OptimizationEnabled` guard somewhere instead of them being scattered throughout, but disabling this folding will have significant CQ impact, so, it is not something for this change I think. Not to mention there is this complicated tradeoff between folding and TP where high-impact folding may lead to an increase in throughput. I am not sure where `fgOptimizeCast` is on that spectrum.  I suppose one thing I know from profiles is that it is pretty cheap.